lang/csharp/src/apache/main/Util/Decimal.cs

```diff
@@ -0,0 +1,177 @@
+
+            var buffer = unscaledInt.ToByteArray();
+
+            Array.Reverse(buffer); // Make the sequence of bytes big-endian
+
+            return Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (object)buffer
+                : (object)new GenericFixed((FixedSchema)schema.BaseSchema, buffer);
+        }
+
+        /// <summary>
+        /// Converts a base value to an instance of the logical type.
+        /// </summary>
+        /// <param name="baseValue">The base value to convert.</param>
+        /// <param name="schema">The schema that represents the target of the conversion.</param>
+        /// <returns>An object representing the encoded value of the logical type.</returns>
+        public override object ConvertToLogicalValue(object baseValue, LogicalSchema schema)
+        {
+            var buffer = Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (byte[])baseValue
+                : ((GenericFixed)baseValue).Value;
+
+            Array.Reverse(buffer); // Make the sequence of bytes little-endian
+
+            var unscaledInt = new BigInteger(buffer);
+            var isNegative = unscaledInt < 0;
+
+            var absUnscaledInt = BigInteger.Abs(unscaledInt);
+
+            buffer = absUnscaledInt.ToByteArray();
+
+            var paddedBuffer = new byte[12];
+
+            Buffer.BlockCopy(buffer, 0, paddedBuffer, 0, Math.Min(buffer.Length, 12));
+
+            var valBits = new int[]
+            {
+                BitConverter.ToInt32(paddedBuffer, 0),
+                BitConverter.ToInt32(paddedBuffer, 4),
+                BitConverter.ToInt32(paddedBuffer, 8)
+            };
+
+            return new decimal(valBits[0], valBits[1], valBits[2], isNegative, (byte)GetLogicalScale(schema));
+        }
+
+        /// <summary>
+        /// Retrieve the .NET type that is represented by the logical type implementation.
+        /// </summary>
+        /// <param name="nullible">A flag indicating whether it should be nullible.</param>
+        public override string GetCSharpTypeName(bool nullible)
+        {
+            var typeName = typeof(decimal).ToString();
+            return nullible ? "System.Nullable<" + typeName + ">" : typeName;
+        }
+
+        /// <summary>
+        /// Determines if a given object is an instance of the logical Decimal.
+        /// </summary>
+        /// <param name="logicalValue">The logical value to test.</param>
+        public override bool IsInstanceOfLogicalType(object logicalValue)
+        {
+            return logicalValue is decimal;
+        }
+
+        private static int GetLogicalScale(LogicalSchema schema)
+        {
+            var scaleVal = schema.GetProperty("scale");
+
+            return string.IsNullOrEmpty(scaleVal) ? 0 : int.Parse(scaleVal);
+        }
+
+        private static byte[] GetUnscaledDecimalValue(decimal value, out int scale)
+        {
+            var buffer = new byte[12];
```

Comment:
Suggest making this `new byte[13]`.  The internals of the decimal are an unsigned integer value, but BigInteger constructor is expecting a signed value. Therefore, easiest way to get BigInteger to interpret the resultant byte[] from this method is to ensure the final bye is 0x00.  This should fix the Decimal.MinValue and Decimal.MaxValue test cases.